Feature market direction

LangLazy opened the pull request:
<!-- Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)
-->
## Summary

Adds new user set optional field which represents market direction, as well as a telegram command allowing users to change this field.

Solve the issue: #8011 

## Quick changelog

- Added new command ` \marketdir <param>` to change the marketdir variable

## What's new?
- Added new field `market_direction `to the `IStrategy `class 
- Added a new enum `MarketDirection`
- Added new command ` \marketdir <param>` to change the marketdir variable
- Added a test to check if the functionality works as intended

The goal of this PR is to allow users to have an internal variable they can control to signify what they think the market conditions are. The default value for this if not set by the user is `MarketDirection.NONE` meaning if a user has not chosen to set up this value, it will not hold any meaning or influence anything else.

LangLazy committed:
initial commit

LangLazy committed:
fixed a test

LangLazy committed:
Merge branch 'freqtrade:develop' into feature

LangLazy committed:
Merge branch 'feature' of github.com:LangLazy/freqtrade into feature

LangLazy committed:
added new text

LangLazy committed:
fixed formatting

LangLazy commented:
Let me know if the unit test and the way I implemented the functionality follow the style of the rest of the code. I would love to get any feedback on what I could do better.

xmatthias commented on freqtrade/rpc/telegram.py:
```diff
@@ -129,7 +129,7 @@ def _init_keyboard(self) -> None:
             r'/forcebuy$', r'/forcelong$', r'/forceshort$',
             r'/forcesell$', r'/forceexit$',
-            r'/edge$', r'/health$', r'/help$', r'/version$'
+            r'/edge$', r'/health$', r'/help$', r'/version$', r'/marketdir$'
```
i think this is pointless.

it'll allow you to create a button `/marketdir` (without anything after it) - but if you call `/marketdir` without argument, it'll do nothing based on the handler, which requires 1 argument.

xmatthias commented on freqtrade/rpc/telegram.py:
```diff
@@ -1677,3 +1678,26 @@ def _send_msg(self, msg: str, parse_mode: str = ParseMode.MARKDOWN,
+        """
+        if context.args and len(context.args) == 1:
+            new_market_dir = context.args[0]
+            match new_market_dir:
```
You can't use match on freqtrade for now (probably not for years).

We do support python 3.8-3.10 - match was only added in 3.10, so this will fail on all prior versions (which tests confirm as "invalid syntax" errors ...

xmatthias commented on freqtrade/rpc/telegram.py:
```diff
@@ -1677,3 +1678,26 @@ def _send_msg(self, msg: str, parse_mode: str = ParseMode.MARKDOWN,
+            new_market_dir = context.args[0]
+            match new_market_dir:
+                case "long":
+                    self._rpc._freqtrade.strategy.market_direction = MarketDirection.LONG
```
conceptually, i'd probably move the actual assignment to the rpc class (which would be mostly aligned how other methods in this regards are setup).

That'll make this (eventually) available easily for other rpc methods (namely, the API) - should we decide to add support for that there as well.
Not a hard requirement - but adding this since you asked :)

xmatthias reviewed (commented):
we'll clearly need some documentation - otherwise it's a pointless feature nobody will know how to use.

LangLazy committed:
addressed some issues mentioned in PR

LangLazy committed:
fixed lint issue

xmatthias commented on freqtrade/rpc/telegram.py:
```diff
@@ -129,7 +129,7 @@ def _init_keyboard(self) -> None:
             r'/forcebuy$', r'/forcelong$', r'/forceshort$',
             r'/forcesell$', r'/forceexit$',
-            r'/edge$', r'/health$', r'/help$', r'/version$'
+            r'/edge$', r'/health$', r'/help$', r'/version$', r'/marketdir \d+$'
```
`\d+` is wrong - as that is regex speak for a number - but we expect `"long", "short", "even", ...`).

Possible solutions will be either generally allowing a string `/marketdir \s+$` - or limiting to allowed strings  `/marketdir (long|short|...)$` (both solutions untested - but most likely somewhat working).

I'd suggest to actually try this (this defines the "valid" entries for [telegram keyboard config](https://www.freqtrade.io/en/stable/telegram-usage/#create-a-custom-keyboard-command-shortcut-buttons)).
You can also use [regex101](https://regex101.com/) to make finding the correct regex easier.

The rest looks fine (from a code perspective) - but is missing docs  ;)

LangLazy replied:
Thanks for the pointers!

I was just a bit confused by what you meant in the last part. Do you suggest implementing the regex you wrote in place of what I have? I didn't quite get what you meant by (not sure what "this" is referencing)

> (this defines the "valid" entries for [telegram keyboard config](https://www.freqtrade.io/en/stable/telegram-usage/#create-a-custom-keyboard-command-shortcut-buttons))

Could you elaborate a bit on what you meant by this?

xmatthias replied:
"this" -> the regex used.

there's a pretty well defined list of acceptable commands - (e.g. `/marketdir long`) - which are currently disallowed (\d+ matches numbers only).
`/marketdir 2` (which the current would allow) is not among the actually working commands ...

LangLazy committed:
Added documentation for new telegram command

LangLazy committed:
fixed command regex and updated documentation

xmatthias reviewed (changes requested):
From a technical perspective, this works quite well. Now it's the details (mainly around usability).

what i think is still missing:
* documentation on how to actually use this within the strategy (i know it's `self.market_direction` - but how will a user know, purely from looking at the docs, without being aware of this functionality / this PR?)
* command feedback
* an entry in the `/help` section of the telegram commands (this helps users looking at it remote only to know WHY the command didn't work)
 
ref 2:
![image](https://user-images.githubusercontent.com/5024695/221374015-46ff4e08-e584-4e50-a6bd-aa0ded0e55cf.png)

a user get's no feedback that the direction was set (misstyping the command has the same effect) -  neither is the currently set direction visible when not passing in an argument (there could be another option for this - but this was the quickest way i could find)

To better explain what i mean:

![image](https://user-images.githubusercontent.com/5024695/221374047-522dab08-01ea-42b6-bbfa-c4e9e792164c.png)

after the 2 above comments - what is market-dir? and how can i tell (as a user)?

LangLazy committed:
Fixed issues raised in PR

LangLazy commented:
Thanks for the feedback!

I have since added command feedback and more descriptive error messages. Let me know if these fix the usability issues. 
Moreover, I added the ability for users to do `/marketdir` which would allow them to find out what market direction is currently set.

I had a couple of questions about your above comments.

I had already added the information about the `/marketdir` command to the `/help` command before you made your last comment, was there somewhere else I needed to update for the `/help` command?

Also regarding

> documentation on how to actually use this within the strategy (i know it's self.market_direction - but how will a user know, purely from looking at the docs, without being aware of this functionality / this PR?)

Where in the documentation should I add that this variable is available for users? I tried looking around but the closest I could find was the page on `strategy-customization` and even then I did not really know where would be appropriate for my change to go.

xmatthias committed:
Improve documentation wording

xmatthias commented:
for now, we'll keep the whole documentation part together in the telegram page.

there's not too much to add other than "use it like this" - and i see this as an edge-case scenario, which very likely not many people will actually use.

LangLazy commented:
Ah, I see, thanks for pointing that out. I think with that last change everything you brought up for review should be addressed. Let me know if anything needs to be updated before this is merged :)

xmatthias reviewed (approved):
LGTM :+1: 

I do see this as beta feature though. If we see this as creating problems (strategies using this will not really produce reproducible results) - or otherwise as problematic, this might be removed in the future.

VietnameseRick commented:
Hi guys, are there any example on how to implement `self.market_direction` to the strategy ?

xmatthias commented:
well usage will be as simple as `self.market_direction == MarketDirection.LONG` (for example).

How you're using that within the context of your strategy is up to your imagination.
Personally, i have my doubts on the benefit of this - as this is something that's set externally (at random, basically), which you can't backtest due to that.

VietnameseRick commented:
> well usage will be as simple as `self.market_direction == MarketDirection.LONG` (for example).
> 
> How you're using that within the context of your strategy is up to your imagination. Personally, i have my doubts on the benefit of this - as this is something that's set externally (at random, basically), which you can't backtest due to that.

I can't share you much details about how I would use it but to be honest, it does help really much, market always change and the bias of trader is what make them out performance compare to bot but `self.market_direction` is what really change the game. Sometime the bias come from the human is better than the indicator

pilchor commented:
Hello,

Does anyone have a working strategy example that uses `self.market_direction`?

When I try to use it, no trades are opened, even though `self.market_direction` is correctly set to `'even'` or `'long'` (as seen in the logs).

Here’s a snippet of my code:

```python
def populate_entry_trend(self, dataframe: DataFrame, metadata: dict) -> DataFrame:
    logger.info(f"Current market direction: {self.market_direction}")
    
    if self.market_direction in ['long', 'even']:
        dataframe.loc[
            (
                (dataframe['close'] > dataframe['senkanA']) &
                (dataframe['close'] > dataframe['senkanB']) &
                (dataframe['close'] > dataframe['tenkan']) &
                (dataframe['senkanB'] > dataframe['senkanA']) &
                (dataframe['tenkan'] > dataframe['kijun'])
            ),
            'enter_long'] = 1

    return dataframe

xmatthias commented:
well market-direction is one part - but you'll still have to have the scenario where all your conditions are true.

Assuming your logic triggers - then it'll set enter_long - but only if **all** conditions within that statement match.
if you struggle to get this running - try with a simpler condition which triggers more frequently? (a condition comparing 5 lines can have days / weeks without trade.
